res/controllers/Pioneer-DDJ-400-script.js

```diff
@@ -0,0 +1,715 @@
+//                 * Jogwheels, Scratching, Bending
+//                 * cycle Temporange
+//                 * Beat Sync
+//                 * Beat Loop Mode
+//                 * Sampler Mode
+//                 * BeatFX (controls Effect Unit 1.  LEFT selects EFFECT1, RIGHT selects EFFECT2, FX_SELECT selects EFFECT3.
+//                   ON/OFF toggles selected effect slot.  SHIFT+ON/OFF disables all three effect slots.
+//                 * Hot Cue Mode
+//
+//             Partially:
+//                 * Beatjump mode (no lighting, shift mode(adjust jump size))
+//                 * PAD FX (only slots A-H, Q-P)
+//                 * Output (lights)
+//                 * Loop Section: Loop in / Out, Call, Double, Half
+//                   (loop adjust not working, '4 beat loop' doesnt work correctly - see comments in PioneerDDJ400.loopin4beatPressedLong)
+//
+//             Testing:
+//                 * Keyboard Mode (check pitch value)
+//                 * Keyshift Mode (check pitch value)
+//
+//             Not working/implemented:
+//                 * Channel & Crossfader Start
+//
+var PioneerDDJ400 = {};
+
+var LightsPioneerDDJ400 = {
+    beatFx: {
+        status: 0x94,
+        data1: 0x47,
+    },
+    deck1: {
+        vuMeter: {
+            status: 0xB0,
+            data1: 0x02,
+        },
+        playPause: {
+            status: 0x90,
+            data1: 0x0B,
+        },
+        shiftPlayPause: {
+            status: 0x90,
+            data1: 0x47,
+        },
+        cue: {
+            status: 0x90,
+            data1: 0x0C,
+        },
+        shiftCue: {
+            status: 0x90,
+            data1: 0x48,
+        },
+    },
+    deck2: {
+        vuMeter: {
+            status: 0xB0,
+            data1: 0x02,
+        },
+        playPause: {
+            status: 0x91,
+            data1: 0x0B,
+        },
+        shiftPlayPause: {
+            status: 0x91,
+            data1: 0x47,
+        },
+        cue: {
+            status: 0x91,
+            data1: 0x0C,
+        },
+        shiftCue: {
+            status: 0x91,
+            data1: 0x48,
+        },
+    },
+};
+
+// Save the Shift State
+PioneerDDJ400.shiftState = [0, 0];
+
+// JogWheel
+PioneerDDJ400.vinylMode = true;
+PioneerDDJ400.alpha = 1.0/8;
+PioneerDDJ400.beta = PioneerDDJ400.alpha/32;
+PioneerDDJ400.highspeedScale = 150; // multiplier for fast seek through track using SHIFT+JOGWHEEL
+PioneerDDJ400.bendScale = 0.5;
+
+PioneerDDJ400.pointJumpSpace = 0.005; // amount in percent of the Song we can jump back to previous Cue or loop point
+
+PioneerDDJ400.tempoRanges = [0.06, 0.10, 0.16, 0.25]; // WIDE = 25%?
+
+// Keyboard Mode Variables and Settings
+PioneerDDJ400.keyboardHotCuePoint = [0, 0]; // selected HotCue point (eg. PAD) in Keyboard mode per Deck 0 = unset
+PioneerDDJ400.keyboardModeRefCount = [0, 0]; // count the currently pressed Pads per Deck
+PioneerDDJ400.halftoneToPadMap = [4, 5, 6, 7, 0, 1, 2, 3];
```

Comment:
ping https://github.com/mixxxdj/mixxx/pull/2312#discussion_r349984342

Reply:
So in looking into this, the reason for the semi-tone mapping is to replicate what rekorbox's does. 

It allows you to jump to a hot cue and adjust the pitch simultaneous. The semi-tones give you a chromatic scale which sounds pretty cool if you jump in time to the beats.

The feature is still pretty experimental here (in contrast to rekordblocks). I'll try polish it up at the end if there's time.

Reply:
Yeah, it's a cool feature. But if you have 8 pads, IMO mapping 8 semitones is odd. Better to map an octave of a scale.

Reply:
@Be-ing I've played around with a few different mappings here.  

* I think chromatic mappings sound best for keyshifting common dance music genres (to very occasionally  spice up a lull for e.g.,).
* I think having the ability to map to a major or (harmonic) minor scale could be useful to play melodies if the track is outputting a constant tone (via e.g., a loop or a track prepared explicitly for this purpose).

One option is to add something like the following:
```js
PioneerDDJ400.chromaticMap  = [4, 5, 6, 7, 0, 1, 2, 3];   // 3/4 chromatic scale
PioneerDDJ400.majorScaleMap = [7, 9, 11, 12, 0, 2, 4, 5];
PioneerDDJ400.minorScaleMap = [7, 8, 11, 12, 0, 2, 3, 5]; // harmonic minor

// Choose your preference for keyboard & keyshift mode from the above 
// (or we could hack a mapping to toggle through them)
PioneerDDJ400.keyboardMap = PioneerDDJ400.chromaticMap;
```

That said, I personally don't find these features very useful either way. :man_shrugging: It sort of feels like they're here "just to be like rekordbox" (where I feel the features kinda suck to begin with).

I'll await your advice on how to proceed, but will add this to the main discussion because it suggests all of the shifted pad modes are better off scrapped...